mathics/builtin/layout.py

```diff
@@ -166,68 +170,34 @@ class Infix(Builtin):
 
     # the right rule should be
     # mbexpression:MakeBoxes[Infix[___], form]
-
-    def eval_infix_1(self, expr: Expression, form: Symbol, evaluation):
-        """MakeBoxes[Infix[expr_],
-        form:StandardForm|TraditionalForm|OutputForm|InputForm]"""
-        return self.do_eval_infix(expr, String("~"), form, evaluation)
-
-    def eval_infix_2(
-        self, expr: Expression, operator: BaseElement, form: Symbol, evaluation
-    ):
-        """MakeBoxes[Infix[expr_, operator_],
-        form:StandardForm|TraditionalForm|OutputForm|InputForm]"""
-        return self.do_eval_infix(expr, operator, form, evaluation)
-
-    def eval_infix_3(
-        self,
-        expr: Expression,
-        operator: BaseElement,
-        precedence: BaseElement,
-        form: Symbol,
-        evaluation,
-    ):
-        """MakeBoxes[Infix[expr_, operator_, precedence_],
+    def eval_all(self, expression, form, evaluation):
+        """MakeBoxes[Infix[___],
         form:StandardForm|TraditionalForm|OutputForm|InputForm]"""
-
+        infix_expr = expression.elements[0]
+        elements = list(infix_expr.elements)
+        num_parms = len(elements)
+
+        if num_parms == 0 or num_parms > 4:
+            evaluation.message("Infix", "argb", Integer(len(parms.get_sequence())))
+            return eval_fullform_makeboxes(infix_expr, evaluation, form)
+
+        if num_parms == 1:
+            expr = elements[0]
+            return self.do_eval_infix(expr, String("~"), form, evaluation)
```

Comment:
These `do_eval` methods in a class really should be  functions outside of any class in a boxing module. 

Note that when I say a boxing module maybe say `mathics.box`.  Note that this kind of "eval" an evaluation that is done in Boxing and we should start making distinct from expression evaluation.

While it is true there may be rare cases where in boxing a general evaluation goes on, for the most part it doesn't.  In other words it is rare that I will want to find to prime integer factors of a number in boxing.  In _formatting_ I might need to compute a value like 3/2 and maybe I feel compelled to use the full strength of WL to do that. But in such rare cases that would be triggered by an explicit `Expression().evaluate()`, rather than implicitly by iterating over an M-Expressions in the boxing process.

Reply:
One other thing I'd like to point out. Note that renaming `self.do_eval_infix` to `mathics.box.box_infix`  is different than collecting the misguided `atom_to_boxes(<form>)` in one place. 

I think there is a subtle difference between  `do_eval_infix` and some of the `atom_to_boxes()` routines. In`do_eval_infix` should be working at a high-level of boxing (in contrast to the atoms which are turning say an image into a base64-encoded PNG wrapped in a box).   

And as such -  working at a high  level - it could make use of form transformation rules (say collected in a file). 

Right now I do not envision testing on the specific form passed to it `if form is ...` but rather that rules that the form object contains are used.   I think this kind of thing was shown in one of your makeboxes implementations in WMA.

Reply:
I just pulled the branch and looked at the code. Maybe gentler than creating a `mathics.box` module for essentially one thing right now (that might be replaced by Mathics rules), would be just to move this out of the class - that is turn it from being a method in a class to a module function - and call it `box_infix()`. 

So what I am also suggesting is that we _not_  use the `eval_` convention for the evaluation aspect of boxing, but instead use a different prefix like `box_`. (Note that "box" is an active verb as well as a noun).

Reply:
What I tried to do here was an small change, keeping the existing structure. But yes, at some point what you suggest is where I am trying to go.

Reply:
To be clear, I think that if the implementation of a `eval_` method is more or less complex, then it should be moved to an external function, in the same module or in another one. I will try to do something like this over this branch, but avoiding changing the code too much. Then, we can try to improve and simplify the internal logic.

Reply:
I totally understands. 

Would pulling out `self.do_eval_infix` out and renaming it to `box_infox` and  `eval_all` to `eval`  be disruptive? 

I note that `self.do_eval_infix()` seems to be new in this branch and is not in master.

Reply:
```Infix.do_eval_infix```  is almost the same code that we have in ```MakeBoxes.eval_infix``` in master.

Reply:
I renamed ```eval_all``` to ```eval_makeboxes``` because I want to be able to find these methods afterward when we can continue with the Format/Box refactor.

Reply:
> `Infix.do_eval_infix` is almost the same code that we have in `MakeBoxes.eval_infix` in master.

Right, so what should eventually happen is that all of those `MakeBoxes.eval_...` methods should be pulled out of Makeboxes and turned into `box_x...` methods. In the future these should not be called by the existing Expression.rewrite_apply_eval()`  but instead called via `MakeBoxes.box(form, ...)`